src/cache-functions.ts

```diff
@@ -8,9 +8,41 @@ import fastFolderSize from 'fast-folder-size';
 import { version } from '../package.json';
 import { extendPaths, extendDependencies } from './env-functions';
 import { isError } from './helper-functions';
+import { parseOptions } from './build-package';
 
 import { EnvironmentVariables } from './types/env-functions';
 
+
+/**
+ * Returns cache key hash
+ * 
+ * @param {string} repo Github repository name
+ * @param {string} cacheSuffix A string which will be appended to the cache key.
+ * @param {EnvironmentVariables} env Local environment object.
+ * @param {string} cmakeOptions Build options string which is added to cache key hash
+ * @param {string} sha Github repository commit SHA
+ * @returns {string} Cache key hash
+ */
+export const getCacheKeyHash = (repo: string, cacheSuffix: string, env: EnvironmentVariables, cmakeOptions: string, sha: string | undefined): string => {
+    const buildOptions = [];
+    buildOptions.push(...parseOptions(cmakeOptions))
+    buildOptions.sort();
+    let cacheKeyStr = `v=${version}${cacheSuffix}::cmake=${env.CMAKE_VERSION}::options=${buildOptions.join()}::${repo}=${sha}`;
+
+    for (const [dependency, dependencySha] of Object.entries(env.DEPENDENCIES || {})) {
+        const [ , dependencyRepo] = dependency.split('/');
+        if (dependencyRepo === repo) continue;
+        cacheKeyStr += `::${dependencyRepo}=${dependencySha}`;
+    }
```

Comment:
Tell me if I'm wrong: I think what we're missing here is the inclusion of the dependency's build options in the cache key string. Or have I missed it?

I think we also want the list of dependencies to be sorted within the cache key. e.g. it shouldn't matter if `metkit` specifies its dependencies to be `eccodes, eckit` or `eckit, eccodes` - the cache key should be the same in either case.

Reply:
Those are very good points, I agree